be/src/runtime/metadata_result_writer.cpp

```diff
@@ -0,0 +1,232 @@
+    }
+    return Status::OK();
+}
+
+void MetadataResultWriter::_init_profile() {
+    _total_timer = ADD_TIMER(_parent_profile, "TotalSendTime");
+    _serialize_timer = ADD_CHILD_TIMER(_parent_profile, "SerializeTime", "TotalSendTime");
+    _sent_rows_counter = ADD_COUNTER(_parent_profile, "NumSentRows", TUnit::UNIT);
+}
+
+Status MetadataResultWriter::append_chunk(Chunk* chunk) {
+    SCOPED_TIMER(_total_timer);
+    auto process_status = _process_chunk(chunk);
+    if (!process_status.ok() || process_status.value() == nullptr) {
+        return process_status.status();
+    }
+    auto result = std::move(process_status.value());
+
+    const size_t num_rows = result->result_batch.rows.size();
+    Status status = _sinker->add_batch(result);
+
+    if (status.ok()) {
+        _written_rows += num_rows;
+        return status;
+    }
+
+    LOG(WARNING) << "Append metadata result to sink failed.";
+    return status;
+}
+
+StatusOr<TFetchDataResultPtrs> MetadataResultWriter::process_chunk(Chunk* chunk) {
+    SCOPED_TIMER(_total_timer);
+    TFetchDataResultPtrs results;
+    auto process_status = _process_chunk(chunk);
+    if (!process_status.ok()) {
+        return process_status.status();
+    }
+    if (process_status.value() != nullptr) {
+        results.push_back(std::move(process_status.value()));
+    }
+    return results;
+}
+
+StatusOr<bool> MetadataResultWriter::try_add_batch(TFetchDataResultPtrs& results) {
+    size_t num_rows = 0;
+    for (const auto& result : results) {
+        num_rows += result->result_batch.rows.size();
+    }
+
+    auto status = _sinker->try_add_batch(results);
+
+    if (status.ok()) {
+        if (status.value()) {
+            _written_rows += num_rows;
+            results.clear();
+        }
+    } else {
+        results.clear();
+        LOG(WARNING) << "Append metadata result to sink failed.";
+    }
+    return status;
+}
+
+StatusOr<TFetchDataResultPtr> MetadataResultWriter::_process_chunk(Chunk* chunk) {
+    if (nullptr == chunk || 0 == chunk->num_rows()) {
+        return nullptr;
+    }
+
+    const int num_columns = _output_expr_ctxs.size();
+
+    Columns result_columns;
+    result_columns.reserve(num_columns);
+
+    for (int i = 0; i < num_columns; ++i) {
+        ASSIGN_OR_RETURN(auto col, _output_expr_ctxs[i]->evaluate(chunk));
+        result_columns.emplace_back(std::move(col));
+    }
+
+    std::unique_ptr<TFetchDataResult> result(new (std::nothrow) TFetchDataResult());
+    if (!result) {
+        return Status::MemoryAllocFailed("memory allocate failed");
+    }
+
+    if (_sink_type == TResultSinkType::METADATA_ICEBERG) {
+        RETURN_IF_ERROR(_fill_iceberg_metadata(result_columns, chunk, result.get()));
+    }
+
+    return result;
+}
+
+// Columns is fixed in the first version of logical iceberg metadata table.
+// In principle, there will be no changes in the future.
+// Only new columns are allowed if necessary. The newly added columns need to check if nullable.
```

Comment:
If add new columns, at upgrage stage, BE version is higher than FE, would get not exist column here?

Reply:
The columns is come from `LogicalIcebergMetadataTable.toThrift`. So the jni scanner can't result the new column. we only need to check if nullable in here is enough.